# Phylo-MIP(Phylogeny-based Molecular Identification Pipeline)
- [README(Êó•Êú¨Ë™û)](README-jp.md)
## What`s app
- This app is a pipeline tool that aims to perform phylogenetic analysis and species determination simply and almost automatically.
- It uses Docker to configure the virtual environment, so it is largely independent of the user's OS or environment.
### Concept
Designed for phylogenetic analysis using environmental DNA and more ...

### What we can do
- Create a FASTA file with OTUs including species names based on the file output by `localBLAST` or `BLAST+` (only for DBs composed of NCBI data) : <b>Phylo-MIP.py</b>
  - Sequence alignment([MAFFT](https://mafft.cbrc.jp/alignment/server/index.html))
  - Haplotype detection([VSEARCH](https://github.com/torognes/vsearch))
  - Creating a phylogenetic tree([FASTTree](https://morgannprice.github.io/fasttree/))
  - Species classification analysis([bPTP](https://species.h-its.org/), [mPTP](https://github.com/Pas-Kapli/mptp))
- Merging the output file of Qiime2 and the output file of Phylo-MIP pipline : <b>merge_data.py</b>
- Confirmed that it can be run on MAcOS or LinuxOS, but it can also be run on WindowsOS using WSL.

## Installation
1. Installing Docker
     - https://docs.docker.jp/engine/getstarted/step_one.html
2. Download or clone Phylo-MIP
  Phylo-MIP is available as a direct download or by cloning the repository.
    <details><summary>DL link</summary>

    ![image](https://github.com/user-attachments/assets/ad10015a-dbe1-4498-a751-ae2e0c42a47a)

    - Download from the `Download ZIP` button
    </details>

    Clone
    ```
    git clone https://github.com/shigebio/Phylo-MIP
    ```

3. Go to DL or cloned file
    ```
    cd /{path to Phylo-MIP}/Phylo-MIP
    ```
4. Creating a virtual environment
    ```
    chmod +x setup.sh
    chmod +x entrypoint.sh
    ./setup.sh
    ```
---
## How to Use Phylo-MIP
**Before you run**
Phylo-MIP.py uses NCBI and GBIF APIs. Please refer to the [NCBI](https://blast.ncbi.nlm.nih.gov/doc/blast-help/developerinfo.html#developerinfo) and [GBIF](https://techdocs.gbif.org/en/openapi/v1/species) API guidelines.
Before running for the first time, as much as possible, please change the example email address in [this code](https://github.com/shigebio/Phylo-MIP/blob/main/app/Phylo-MIP.py#L22) to your own.

Part of the NCBI guidelines
>- Do not contact the server more often than once every 10 seconds.
>- Do not poll for any single RID more often than once a minute.
>- Use the URL parameter email and tool, so that the NCBI can contact you if there is a problem.
>- Run scripts weekends or between 9 pm and 5 am Eastern time on weekdays if more than 50 searches will be submitted.
---

1. Executing commands

    **Basic commands**
      ```
      phylo-mip {Path to Input CSV} --tree {Options}
      ```
    **Example**
      ```
      phylo-mip ./path/to/your_input.csv --tree --method ML --bootstrap 250
      ```

   See [here](https://github.com/shigebio/Phylo-MIP/blob/main/README-Preparing_the_input_files.md) for acceptable input file formats

      <details><summary>See Options</summary>

        - `--top` : You can specify 1 to 10 of the top `pident` with the same `qseqid`.
          - Default: `1`
        - `--o` : Output name of the output file. If not present, the default name is applied.
        - `--class` : Analysis can be limited to specific taxa (currently only classes).
          - Please note that if taxonomic group information obtained from NCBI or GBIF is not entered, it will not be searchable.
          - You can also manually create a `class` column and enter any value.
        - `--tree`
            ```
            ## Sub options
            # --method : You can select the method for constructing a phylogenetic tree. Only `ML` can be used, which will use the ML method.
            # --bootstrap : Number of bootstrap iterations. Default: `250`
            # --gamma : Whether to apply gamma distribution. Default: `False`
            # --outgroup {OTU name} : Specify the outgroup
        - `--onlyp`Ôºö Perform phylogenetic analysis and beyond (alignment ‚Üí removal of identical haplotypes ‚Üí creation of phylogenetic tree ‚Üí species determination analysis)
            - https://github.com/shigebio/MICUM/pull/7
          - `--bptp` : bPTP Analysis Options
            ```
            ## Sub options
            # --mcmc : You can set the number of MCMC iterations.
            # --thinning : You can set the number of samplings. Sampling is performed for each MCMC chain every specified number. Default: `100`
            # --burnin : Burn-in ratio (0.1~1.0). Default: `0.1`
            # --seed : The seed you want to assign. The same seed value will give you the same results every time for the same input. The default is a random seed.
      </details>

   **If you only want to output FASTA and CSV files**
      ```
      phylo-mip {input CSV file name} {output file name}
      ```
      **Example**
      ```
      phylo-mip your_data.csv output
      ```

## Outputs
```
=== Directory Structure Verification ===
micum_output_{ÂÆüË°åÊôÇÂàª}/
  taxonomy/ # ÂàÜÈ°ûÊÉÖÂ†±„Çí‰ªò‰∏é„Åó„ÅüCSV„ÄÅFASTA„Éï„Ç°„Ç§„É´
    taxonomic_data.csv
    taxonomic_sequences.fasta
  phylogeny/ # fattree„Å´„Çà„ÇãÁ≥ªÁµ±Ê®π
    {ÂÆüË°åÊôÇÈñì}_ML_phylogenetic_tree.nex # „Éï„Ç°„Ç§„É´„ÅÆÈ†≠„Å´„ÅØÂÆüË°åÊôÇ„ÅÆprefix„Åå‰ªò„Åç„Åæ„Åô
    {ÂÆüË°åÊôÇÈñì}_ML_{ÂÆüË°åÊôÇÈñì}_ML_phylogenetic_tree.nwk
  alignment/ # mafft„ÄÅVSEARCHÂÆüÊñΩÂæå„ÅÆ„Éï„Ç°„Ç§„É´
    {ÂÆüË°åÊôÇÈñì}_haplotype_clusters.tsv # VSEARCH„Å´„Çà„Çã„Éè„Éó„É≠„Çø„Ç§„ÉóÂêåÂÆöÁµêÊûúTSV„Éï„Ç°„Ç§„É´
    {ÂÆüË°åÊôÇÈñì}_haplotype_clusters.csv„ÄÄ# VSEARCH„Å´„Çà„Çã„Éè„Éó„É≠„Çø„Ç§„ÉóÂêåÂÆöÁµêÊûúCSV„Éï„Ç°„Ç§„É´
    {ÂÆüË°åÊôÇÈñì}_clustered_sequences.fasta # VSEARCH„Å´„Çà„ÇãÂêå‰∏Ä„Éè„Éó„É≠„Çø„Ç§„Éó„ÅÆÈô§ÂéªÂæå„ÅÆ„Éï„Ç°„Ç§„É´
    {ÂÆüË°åÊôÇÈñì}_aligned_sequences.fasta # MAFFT„Å´„Çà„Çã„Ç¢„É©„Ç§„É°„É≥„ÉàÂæå„ÅÆ„Éï„Ç°„Ç§„É´
  mptp/ # mptpÂá∫ÂäõÁµêÊûú
    {ÂÆüË°åÊôÇÈñì}_mPTP_analysis/
      {ÂÆüË°åÊôÇÈñì}_mPTP_species_delimitation.txt # Ââ≤„ÇäÂΩì„Å¶„Çâ„Çå„ÅüSpecies„Å®ÂêÑSpecies„ÅÆÂ∞§Â∫¶„ÇíÂê´„ÇÄTXT„Éï„Ç°„Ç§„É´
      {ÂÆüË°åÊôÇÈñì}_mPTP_species_delimitation.svg # Ê®πÂΩ¢„Éï„Ç°„Ç§„É´(SVG)
  bptp/ # bptpÂá∫ÂäõÁµêÊûú
    {ÂÆüË°åÊôÇÈñì}_bPTP_analysis/
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.PTPhSupportPartition.txt.svg # Á∞°Êòì„Éí„É•„Éº„É™„Çπ„ÉÜ„Ç£„ÉÉ„ÇØÊ§úÁ¥¢(simple heuristic search)„Å´„Çà„ÇäÊßãÁØâ„Åï„Çå„ÅüÊ®πÂΩ¢„Éï„Ç°„Ç§„É´(SVG)
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.llh.pdf # MCMC chain„ÅÆtrace„É≠„Ç∞„Ç∞„É©„Éï(PDF)
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.PTPhSupportPartition.txt.sh.tre # Á∞°Êòì„Éí„É•„Éº„É™„Çπ„ÉÜ„Ç£„ÉÉ„ÇØÊ§úÁ¥¢(simple heuristic search)„Å´„Çà„ÇäÊßãÁØâ„Åï„Çå„ÅüÊ®πÂΩ¢„Éï„Ç°„Ç§„É´(TREE)
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.PTPMLPartition.txt # MLÊ≥ï„Å´„Çà„ÇäÂâ≤„ÇäÂΩì„Å¶„Çâ„Çå„ÅüSpecies„Å®ÂêÑSpecies„ÅÆÂ∞§Â∫¶„ÇíÂê´„ÇÄTXT„Éï„Ç°„Ç§„É´
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.PTPMLPartition.txt.ml.tre # MLÊ≥ï„Å´„Çà„ÇäÊßãÁØâ„Åï„Çå„ÅüÊ®πÂΩ¢„Éï„Ç°„Ç§„É´(TREE)
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.PTPMLPartition.txt.png # MLÊ≥ï„Å´„Çà„ÇäÊßãÁØâ„Åï„Çå„ÅüÊ®πÂΩ¢„Éï„Ç°„Ç§„É´(PNG)
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.PTPllh.txt # MCMC chain„ÅÆtrace„É≠„Ç∞
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.PTPhSupportPartition.txt.png # Á∞°Êòì„Éí„É•„Éº„É™„Çπ„ÉÜ„Ç£„ÉÉ„ÇØÊ§úÁ¥¢(simple heuristic search)„Å´„Çà„ÇäÊßãÁØâ„Åï„Çå„ÅüÊ®πÂΩ¢„Éï„Ç°„Ç§„É´(PNG)
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.PTPPartitions.txt
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.PTPhSupportPartition.txt # Á∞°Êòì„Éí„É•„Éº„É™„Çπ„ÉÜ„Ç£„ÉÉ„ÇØÊ§úÁ¥¢(simple heuristic search)„Å´„Çà„ÇäÂâ≤„ÇäÂΩì„Å¶„Çâ„Çå„ÅüSpecies„Å®ÂêÑSpecies„ÅÆÂ∞§Â∫¶„ÇíÂê´„ÇÄTXT„Éï„Ç°„Ç§„É´
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.PTPMLPartition.txt.svg # MLÊ≥ï„Å´„Çà„ÇäÊßãÁØâ„Åï„Çå„ÅüÊ®πÂΩ¢„Éï„Ç°„Ç§„É´(SVG)
      {ÂÆüË°åÊôÇÈñì}_bPTP_species_delimitation.PTPPartitonSummary.txt # ÂêÑÊâãÊ≥ï„ÅßÂàÜÂâ≤„Åï„Çå„ÅüSpeiceis
```

The taxonomic status of sequences assigned to an OTU is assigned according to the value of pident (match rate of search sequence in localBLAST search) as follows:
```
pident >= 98.00 : Species name
95.00 <= pident < 98.00 : Genus name
90.00 <= pident < 95.00 : Family name
85.00 <= pident < 90.00 : Order name
```
code: https://github.com/shigebio/Phylo-MIP/blob/main/app/Phylo-MIP.py#L291-L298

<b>ÔºäCAUTIONÔºä
     Due to the structure of the NCBI database, the taxon information columns may be out of sync when retrieved from the Entrez API (NCBI). Please check the taxon information in the file.
</b>

---
## How to Use merge_data
**Basic commands**
  ```
  merge_data -q {Qiime output file path} -m {Phylo-MIP pipeline output file path} -f {The file format you want to output: csv/tsv} -o {output file name}
  ```

Output file name default: time plefix on executed.

The merged file will be output to woking directory.

# The problems you have experienced
‚ÜíPlease [create a new issue](https://github.com/shigebio/Phylo-MIP/issues) and include the detailsüôè
